transforms/pair.go

```diff
@@ -0,0 +1,137 @@
+package transforms
+
+import (
+	"github.com/facebookgo/errgroup"
+	"gopkg.in/Clever/optimus.v3"
+)
+
+// PairType is the type of join to use when Pairing
+type PairType int
+
+// RowHasher takes in a row and returns a hash for that Row.
+// Used when Pairing.
+type RowHasher func(optimus.Row) (interface{}, error)
+
+// KeyHasher is a convenience function that returns a RowHasher that hashes based on the value of a
+// key in the Row.
+func KeyHasher(key string) RowHasher {
+	return func(row optimus.Row) (interface{}, error) {
+		return row[key], nil
+	}
+}
+
+const (
+	// LeftJoin keeps any row where a Row was found in the left Table.
+	LeftJoin PairType = iota
+	// RightJoin keeps any row where a Row was found in the right Table.
+	RightJoin
+	// InnerJoin keeps any row where a Row was found in both Tables.
+	InnerJoin
+	// OuterJoin keeps all rows.
+	OuterJoin
+)
+
+// mustHave takes in any amount of keys and returns a function that can be passed to Select
+// that returns true for any Row that has all of those keys.
+func mustHave(keys ...string) func(optimus.Row) (bool, error) {
+	return func(row optimus.Row) (bool, error) {
+		for _, key := range keys {
+			if row[key] == nil {
+				return false, nil
+			}
+		}
+		return true, nil
+	}
+}
+
+// Pair returns a TransformFunc that pairs all the elements in the table with another table, based
+// on the given hashing functions and join type.
+func Pair(rightTable optimus.Table, leftHash, rightHash RowHasher, join PairType) optimus.TransformFunc {
```

Comment:
This definitely isn't something for this change, but it might be worth thinking about a separate join interface in v4 that was separate from TransformFunc. That way we could do `joinedTable := Join(leftTable, rightTable, leftJoinAttribute, rightJoinAttribute, JoinType)`